File: AdvancedTypes/src/main/java/com/bennyhuo/kotlin/advancedtypes/dataclasses/java/Book.java
package com.bennyhuo.kotlin.advancedtypes.dataclasses.java;

public class Book {
    private long id;
    private String name;
    private Person person;

    public long getId() {
        return id;
    }

    public Book id(long id) {
        this.id = id;
        return this;
    }

    public String getName() {
        return name;
    }

    public Book name(String name) {
        this.name = name;
        return this;
    }

    public Person getPerson() {
        return person;
    }

    public Book person(Person person) {
        this.person = person;
        return this;
    }
}

File: AdvancedTypes/src/main/java/com/bennyhuo/kotlin/advancedtypes/dataclasses/kotlin/Book.kt
package com.bennyhuo.kotlin.advancedtypes.dataclasses.kotlin

import com.bennyhuo.kotlin.advancedtypes.dataclasses.PoKo

@PoKo
data class Book(val id: Long,
                val name: String,
                val author: Person) {

}

data class Person(val id: Long, val name: String, val age: Int)

fun main() {
//    val book = Book(0, "Kotlin in Action", Person(1, "Dmitry", 40))
//    book.copy()
////    val id = book.component1()
////    val name = book.component2()
////    val author = book.component3()
//
//    val (id, name, author) = book
//
//    val pair = "Hello" to "World"
//    val (hello, world) = pair

    val book = Book::class.java.newInstance()

}